### PR TITLE
Add PL metrics for scalar models

### PR DESCRIPTION
**Pull request by melanibe**
Write a detailed description of your PR here.

Please follow the guidelines for PRs contained [here](docs/pull_requests.md). Checklist:

- [ ] Ensure that your PR is small, and implements one change.
- [ ] Add unit tests for all functions that you introduced or modified.
- [ ] Run PyCharm's code cleanup tools on your Python files.
- [ ] Link the correct GitHub issue for tracking.
- [ ] Update the [Changelog](CHANGELOG.md) file: Describe your change in terms of 
Added/Changed/Removed/... in the "Upcoming" section.
- [ ] When merging your PR, replace the default merge message with a description of your PR,
and if needed a motivation why that change was required.

### CODE REVIEW

File: InnerEye/ML/lightning_models.py [hunk 1]
```diff
@@ -412,18 +423,38 @@ def __init__(self, config: ScalarModelBase, *args, **kwargs) -> None:
...
+                    AreaUnderRocCurve(),
+                    AreaUnderPRCurve(),
+                    BinaryCrossEntropy()]
+        else:
+            return [MeanAbsoluteError(), MeanSquaredError(), ExplainedVariance()]
+
+    def create_metrics(self):
+        self.train_metrics_dict = OrderedDict(
+            [(p, self._get_metrics_classes()) for p in self.sequence_target_positions])
+        self.val_metrics_dict = OrderedDict([(p, self._get_metrics_classes()) for p in self.sequence_target_positions])
```
Comment (ant0nsc) [inline InnerEye/ML/lightning_models.py:456]: Are those fields also initialized (and typed) in the constructor?

File: InnerEye/ML/lightning_models.py [hunk 2]
```diff
@@ -412,18 +423,38 @@ def __init__(self, config: ScalarModelBase, *args, **kwargs) -> None:
...
+                    FalseNegativeRateOptimalThreshold(),
+                    AreaUnderRocCurve(),
+                    AreaUnderPRCurve(),
+                    BinaryCrossEntropy()]
+        else:
+            return [MeanAbsoluteError(), MeanSquaredError(), ExplainedVariance()]
+
+    def create_metrics(self):
+        self.train_metrics_dict = OrderedDict(
+            [(p, self._get_metrics_classes()) for p in self.sequence_target_positions])
```
Comment (ant0nsc) [inline InnerEye/ML/lightning_models.py:455]: there's a problem here I think - you are using sequence target position as the key, but later you use them as if they were hue names.

File: InnerEye/ML/lightning_models.py [hunk 3]
```diff
@@ -442,22 +481,33 @@ def training_or_validation_step(self,
...
+            # mask the model outputs and labels if required
+            masked_model_outputs_and_labels = get_masked_model_outputs_and_labels(
+                logits[:, i, ...], targets[:, i, ...])
+            # compute metrics on valid masked tensors only
+            if masked_model_outputs_and_labels is not None:
+                _model_output, _labels = \
+                    masked_model_outputs_and_labels.model_outputs.data, \
+                    masked_model_outputs_and_labels.labels.data
+                # Convert labels to the same datatype as the model outputs, necessary when running with AMP
+                _labels = _labels.to(dtype=_model_output.dtype)
```
Comment (ant0nsc) [inline InnerEye/ML/lightning_models.py:505]: do we still need that? We've gotten rid of a lot of the explicit conversion code, and left that all to PL.

File: InnerEye/ML/metrics.py [hunk 4]
```diff
@@ -32,6 +36,140 @@
...
+
+class Accuracy05(metrics.Accuracy):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.name = MetricType.ACCURACY_AT_THRESHOLD_05.value
+
+class ScalarMetricsBase(Metric):
+    def __init__(self, dist_sync_on_step: bool = False, name: str = "metric"):
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+        self.add_state("preds", default=[], dist_reduce_fx=None)
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:65]: preds -> predictions? Or use a name that is consistent with the names that PL metrics classes use

File: InnerEye/ML/metrics.py [hunk 5]
```diff
@@ -32,6 +36,140 @@
...
+        self.name = MetricType.EXPLAINED_VAR.value
+
+
+class Accuracy05(metrics.Accuracy):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.name = MetricType.ACCURACY_AT_THRESHOLD_05.value
+
+class ScalarMetricsBase(Metric):
+    def __init__(self, dist_sync_on_step: bool = False, name: str = "metric"):
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:63]: I'd use "" as the default value for `name`

File: InnerEye/ML/metrics.py [hunk 6]
```diff
@@ -32,6 +36,140 @@
...
+    def compute(self):
+        raise NotImplementedError("Should be implemented in the child classes")
+
+    def _get_metrics_at_optimal_cutoff(self, preds, targets) -> Tuple:
+        """
+        Computes the ROC to find the optimal cut-off i.e. the probability threshold for which the
+        difference between true positive rate and false positive rate is smallest. Then, computes
+        the false positive rate, false negative rate and accuracy at this threshold (i.e. when the
+        predicted probability is higher than the threshold the predicted label is 1 otherwise 0).
+        :param hue: The hue to restrict the values used for computation, otherwise all values will be used.
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:82]: outdated doc,  no such param anymore

File: InnerEye/ML/metrics.py [hunk 7]
```diff
@@ -32,6 +36,140 @@
...
+class AreaUnderRocCurve(ScalarMetricsBase):
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.AREA_UNDER_ROC_CURVE.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        targets = torch.cat(self.targets)
+        if torch.unique(targets).numel() == 1:
+            return torch.tensor(np.nan)
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:147]: is there a specific reason for using nan here? Did we do that in the old code too? If not, how about returning 0 if prediction is not correct, and 1 otherwise

File: InnerEye/ML/metrics.py [hunk 8]
```diff
@@ -32,6 +36,140 @@
...
+        return auroc(preds, targets)
+
+
+
+class AreaUnderPRCurve(ScalarMetricsBase):
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:152]: Maybe spell out? AreaUnderPrecisionRecallCurve?

File: InnerEye/ML/metrics.py [hunk 9]
```diff
@@ -32,6 +36,140 @@
...
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step,
+                         name=MetricType.AREA_UNDER_PR_CURVE.value)
+
+    def compute(self):
+        preds = torch.cat(self.preds)
+        targets = torch.cat(self.targets)
+        if torch.unique(targets).numel() == 1:
+            return torch.tensor(np.nan)
```
Comment (ant0nsc) [inline InnerEye/ML/metrics.py:161]: Same Q as above, why nan?

File: Tests/ML/test_metrics.py [hunk 10]
```diff
@@ -154,3 +159,68 @@ def test_get_column_name_for_logging() -> None:
...
+    output = [torch.tensor([0.9, 0.8, 0.6]), torch.tensor([0.3, 0.9, 0.4])]
+    label = [torch.tensor([1., 1., 0.]), torch.tensor([0., 0., 0.])]
+    n_batches = len(output)
+    for i in range(n_batches):
+        metrics.update(output[i], label[i])
+    result = metrics.compute()
+    assert result == 4.0 / 6
+
+def test_classification_metrics():
+    metrics = [AccuracyAtOptimalThreshold(), OptimalThreshold(),
```
Comment (ant0nsc) [inline Tests/ML/test_metrics.py:176]: can you rely on _get_metrics_classes here, rather than re-defininig the list of metrics that you are using?

File: Tests/ML/test_metrics.py [hunk 11]
```diff
@@ -154,3 +159,68 @@ def test_get_column_name_for_logging() -> None:
...
+               FalsePositiveRateOptimalThreshold(), FalseNegativeRateOptimalThreshold(),
+               AreaUnderRocCurve(), AreaUnderPRCurve(), BinaryCrossEntropy()]
+    output = [torch.tensor([0.9, 0.8, 0.6]), torch.tensor([0.3, 0.9, 0.4])]
+    label = [torch.tensor([1., 1., 0.]), torch.tensor([0., 0., 0.])]
+    n_batches = len(output)
+    for i in range(n_batches):
+        for metric in metrics:
+            metric.update(output[i], label[i])
```
Comment (ant0nsc) [inline Tests/ML/test_metrics.py:184]: indexing is error prone, how about using zip(output, label)?

File: InnerEye/ML/lightning_models.py (hunk 2 above)
Comment (ant0nsc) [reply InnerEye/ML/lightning_models.py:455]: On that note, is there a test that ensures that the loggers see the correct metric names, including prediction targets? There is a simple StoringLogger already that you could use for that.

**Review by ant0nsc — CHANGES_REQUESTED** [review]

File: InnerEye/ML/metrics.py (hunk 4 above)
Comment (melanibe) [reply InnerEye/ML/metrics.py:65]: They actually use this "preds" and "target" that's why i chose these names

File: InnerEye/ML/metrics.py (hunk 7 above)
Comment (melanibe) [reply InnerEye/ML/metrics.py:147]: So sklearn pr_curve (which we used previously) returns a warning saying "hey what you are computing does not make sense you have only one class". But the PL version raises an error when this occurs. Since the metrics does not make sense anyway I thought it may be safer to return nan than a value that does not really have a meaning

File: InnerEye/Common/metrics_dict.py [hunk 12]
```diff
@@ -99,7 +99,7 @@ class MetricType(Enum):
     # Regression metrics
     MEAN_ABSOLUTE_ERROR = "MeanAbsoluteError"
     MEAN_SQUARED_ERROR = "MeanSquaredError"
-    R2_SCORE = "r2Score"
+    EXPLAINED_VAR = "ExplainedVariance"
```
[the remark this answers is not in the document]
Comment (melanibe) [reply InnerEye/Common/metrics_dict.py:102]: Don't think so

File: InnerEye/ML/lightning_models.py (hunk 2 above)
Comment (melanibe) [reply InnerEye/ML/lightning_models.py:455]: So I wanted to get rid of the hue name since it was just "hue + sequence_target". Here self.sequence_target_positions didn't seem to used anywhere anymore anyway, hence i "re-purposed" it. But I will rename it to `self.target_names` to avoid confusion

File: InnerEye/ML/lightning_models.py (hunk 3 above)
Comment (melanibe) [reply InnerEye/ML/lightning_models.py:505]: Good question...

File: InnerEye/ML/lightning_models.py (hunk 1 above)
Comment (melanibe) [reply InnerEye/ML/lightning_models.py:456]: Updated

File: InnerEye/ML/lightning_models.py (hunk 3 above)
Comment (melanibe) [reply InnerEye/ML/lightning_models.py:505]: Nope

File: InnerEye/ML/metrics.py (hunk 7 above)
Comment (ant0nsc) [reply InnerEye/ML/metrics.py:152]: I see, yes, here we'd only see NaN if the whole, say, validation set is a single element.

File: InnerEye/ML/lightning_models.py [hunk 13]
```diff
@@ -442,22 +481,31 @@ def training_or_validation_step(self,
...
-            self.log_on_epoch(name=name + hue_suffix, value=value, is_training=is_training)
+        self.compute_and_log_metrics(logits, labels, is_training)
         self.log_on_epoch(name=MetricType.SUBJECT_COUNT,
                           value=len(model_inputs_and_labels.subject_ids),
                           is_training=is_training,
                           reduce_fx=sum)
+        self.log_time_per_batch(is_training)
         return loss
 
+    def compute_and_log_metrics(self, logits: torch.Tensor, targets: torch.Tensor, is_training: bool):
```
Comment (ant0nsc) [inline InnerEye/ML/lightning_models.py:492]: Could you return here please also a list of tuples (subject, prediction_target, model_output, label)?
